File: fft/Fft.java
// Fast Fourier Transform

public class Fft {
    public static final float TAU = (float) (2*Math.PI);
    static class Complex {
        public float re;
        public float im;

        public Complex(float real, float imaginary) {
            this.re = real;
            this.im = imaginary;
        }

        public String toString() {
            return "(" + this.re + ", " + this.im + ")";
        }

        public Complex myclone() {
            return new Complex(this.re, this.im);
        }

        public Complex add(Complex c) {
            this.re += c.re;
            this.im += c.im;
            return this;
        }

        public Complex sub(Complex c) {
            this.re -= c.re;
            this.im -= c.im;
            return this;
        }

        public Complex mul(Complex c) {
            float r = this.re * c.re - this.im * c.im;
            float i = this.re * c.im + this.im * c.re;
            this.re = r;
            this.im = i;
            return this;
        }

        public Complex mul(float f) {
            this.re *= f;
            this.im *= f;
            return this;
        }

        public static Complex add(Complex a, Complex b) {
            return a.myclone().add(b);
        }

        public static Complex sub(Complex a, Complex b) {
            return a.myclone().sub(b);
        }

        public static Complex mul(Complex a, Complex b) {
            return a.myclone().mul(b);
        }

        public static Complex expi(float theta) {
            return new Complex((float) Math.cos(theta), (float) Math.sin(theta));
        }
    }

    public static int c1 = 0;
    public static int c2 = 0;

    public static Complex[] FFT(Complex[] P, boolean inverse) {
        c1++;
        final int n = P.length;
        final int n2 = n/2;
        if (n == 1) {
            return P;
        }
        final float sign = inverse ? -1f : 1f;
        final Complex w = Complex.expi(sign*TAU/n);
        Complex[] Pe = new Complex[n2];
        Complex[] Po = new Complex[n2];
        for (int i = 0; i < n2; i++) {
            Pe[i] = P[2*i];
            Po[i] = P[2*i+1];
        }
        Complex[] ye = FFT(Pe, inverse);
        Complex[] yo = FFT(Po, inverse);
        Complex[] y = new Complex[n];
        Complex wk = new Complex(1f, 0f);
        for (int j = 0; j < n2; j++) {
            c2++;
            Complex yowk = Complex.mul(yo[j], wk);
            y[j] = Complex.add(ye[j], yowk);
            y[j + n2] = Complex.sub(ye[j], yowk);
            wk.mul(w);
        }
        return y;
    }

    public static Complex[] IFFT(Complex[] P) {
        Complex[] res = FFT(P, true);
        for (int i = 0; i < P.length; i++)
            res[i].mul((float)(1. / P.length));
        return res;
    }

    public static void printr(Object[] o) {
        System.out.println(java.util.Arrays.toString(o));
    }

    public static Complex[] MUL(Complex[] A, Complex[] B) {
        int n = 1<<(32 - Integer.numberOfLeadingZeros(A.length + B.length - 2));
        Complex[] a = new Complex[n];
        Complex[] b = new Complex[n];
        System.arraycopy(A, 0, a, 0, A.length);
        System.arraycopy(B, 0, b, 0, B.length);
        for (int i = A.length; i < n; i++)
            a[i] = new Complex(0, 0);
        for (int i = B.length; i < n; i++)
            b[i] = new Complex(0, 0);
        Complex[] X = FFT(a, false);
        Complex[] Y = FFT(b, false);
        Complex[] Z = new Complex[n];
        for (int i = 0; i < n; i++)
            Z[i] = Complex.mul(X[i], Y[i]);
        Complex[] C = IFFT(Z);
        return C;
    }

    public static void main(String[] args) {
        int N = 140000;
        Complex[] A = new Complex[N];
        Complex[] B = new Complex[N];
        for (int i = 0; i < N; i++) {
            A[i] = new Complex(i + 1, 0);
            B[i] = new Complex(1400 - i, 0);
        }

        //import datetime
        //start = datetime.datetime.now()
        for (int i = 0; i < 1; i++) {
            MUL(A, B);
        }
        System.out.println(c1);
        System.out.println(c2);
        //elapsed = datetime.datetime.now() - start;
        //print(elapsed)


        //Complex[] X = {new Complex(1, 0), new Complex(-1, 0), new Complex(2, 0), new Complex(1, 0)};
        //Complex[] Y = {new Complex(-3, 0), new Complex(2, 0), new Complex(-4, 0), new Complex(2, 0)};
        //let C = [-3, 5, -12, 7, -8, 0, 2]
        //System.out.println(java.util.Arrays.toString(MUL(X, Y)));
    }
}

File: fft/fft.js
// Fast Fourier Transform

class Complex {
    re; im;

    constructor(real, imaginary) {
        this.re = real;
        this.im = imaginary;
    }

    toString() {
        return "(" + this.re + ", " + this.im + ")";
    }

    clone() {
        return new Complex(this.re, this.im);
    }

    add(c) {
        this.re += c.re;
        this.im += c.im;
        return this;
    }

    sub(c) {
        this.re -= c.re;
        this.im -= c.im;
        return this;
    }

    mul(c) {
        let r = this.re * c.re - this.im * c.im;
        let i = this.re * c.im + this.im * c.re;
        this.re = r;
        this.im = i;
        return this;
    }

    static add(a, b) {
        return a.clone().add(b);
    }

    static sub(a, b) {
        return a.clone().sub(b);
    }

    static mul(a, b) {
        return a.clone().mul(b);
    }

    static expi(theta) {
        return new Complex(Math.cos(theta), Math.sin(theta));
    }
}


function FFT(P, inverse) {
    const n = P.length
    if (n == 1) {
        return typeof P[0] === "object" ? P : [new Complex(P[0], 0)]
    }
    const sign = inverse ? -1 : 1
    const w = Complex.expi(sign*2*Math.PI/n)
    let Pe = new Array(n>>1);
    let Po = new Array(n>>1);
    for (let i = 0; i < n/2; i++) {
        Pe[i] = P[2*i];
        Po[i] = P[2*i+1];
    }
    const ye = FFT(Pe, inverse)
    const yo = FFT(Po, inverse)
    let y = new Array(n)
    let wk = new Complex(1, 0)
    for (let j = 0; j < (n>>1); j++) {
        y[j] = Complex.add(ye[j], Complex.mul(wk, yo[j]))
        y[j + (n>>1)] = Complex.sub(ye[j], Complex.mul(wk, yo[j]))
        wk.mul(w)
    }
    return y
}

function IFFT(P) {
    return FFT(P, true).map(x => x.re / P.length);
}

function bitCount(n) {
    let count = 0;
    while (n != 0){
        count++;
        n >>= 1;
    }
    return count;
}

function MUL(A, B) {
    let n = 1<<bitCount(A.length + B.length - 2)
    let a  = A.concat(new Array(n - A.length).fill(0))
    let b  = B.concat(new Array(n - B.length).fill(0))
    let X = FFT(a)
    let Y = FFT(b)
    let Z = new Array(n)
    for (let i = 0; i < n; i++)
        Z[i] = Complex.mul(X[i], Y[i])
    let C = IFFT(Z)
    while (n --> 0 && C[n] == 0)
        C.splice(-1, 1)
    return C.map(x => Math.round(x))
}

/*
let A = [1, -1, 2, 1]
let B = [-3, 2, -4, 2]
let C = [-3, 5, -12, 7, -8, 0, 2]
console.log(C)
console.log(MUL(A, B))
*/

let N = 140000
let A = new Array(N)
let B = new Array(N)
for (let i = 0; i < N; i++) {
    A[i] = i + 1
    B[i] = 1400 - i
}

//import datetime
//start = datetime.datetime.now()
for (let i = 0; i < 1; i++)
    MUL(A, B)
//elapsed = datetime.datetime.now() - start;
//print(elapsed)

/*
#C = MUL(A, B)
#print(C)
#print("Size: ", len(C))
*/
